File: Pre-Written Code/Data Structures/2D Fenwick Tree.cpp
#include <iostream>
#include <vector>
using namespace std;
typedef long long ll;

//Queries take O(log^2 n) time
struct FT2D {
	int n, m; vector<vector<int>> vals; FT2D() {}
	FT2D(int _n, int _m) : n(_n), m(_m), vals(_n + 1, vector<int>(_m + 1, 0)) {}

	//Adds val to (i,j) 0-indexed
	void update(int i, int j, int val) {
		for (++i, ++j; i <= n; i += i & -i)
			for (int _j = j; _j <= m; _j += _j & -_j) 
				vals[i][_j] += val;
	}

	//Sum of the rectangle (0,0), (i,j)  0-indexed
	int query(int i, int j) const {
		int sum = 0;
		for (++i, ++j; i; i -= i & -i)
			for (int _j = j; _j; _j -= _j & -_j)
				sum += vals[i][_j];
		return sum;
	}

	//Sum of the rectangle (i1,j1), (i2,j2)
	int query(int i1, int j1, int i2, int j2) const {
		return query(i2, j2) - query(i2, j1 - 1) - 
			query(i1 - 1, j2) + query(i1 - 1, j1 - 1);
	}
};

int main() {
	ios::sync_with_stdio(0);
	cin.tie(0), cout.tie(0);

}

File: Pre-Written Code (Python)/Data Structures/2D Fenwick Tree.py
# Queries take O(log^2 n) time
class FT2D:
    def __init__(self, n, m):
        self.n = n
        self.m = m
        self.vals = [[0]*(m+1) for i in range(n+1)]
    # Adds val to (i,j) 0-indexed
    def update(self, i, j, val):
        i += 1
        j += 1
        while i<=self.n:
            _j = j
            while _j <= self.m:
                self.vals[i][_j] += val
                _j += _j & -j
            i += i & -i                
    def query(self, i, j, i2='', j2=''):
        if isinstance(i2, str):
            # Sum of the rectangle (0,0),(i,j) 0-indexed
            sum = 0
            i += 1
            j += 1
            while i:
                _j = j
                while _j:
                    sum += self.vals[i][_j]
                    _j -= _j & -_j
                i -= i & -i
            return sum
        else:
            # Sum of the rectangle (i1, j1, i2, j2)
            return self.query(i2, j2) - self.query(i2,j-1) - self.query(i-1, j2) + self.query(i - 1, j - 1)
